rewrite-core/src/main/java/org/openrewrite/text/CreateTextFile.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.text;
+
+import org.openrewrite.*;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+
+
+public class CreateTextFile extends Recipe {
+
+    @Option(displayName = "File Contents",
+            description = "Multiline text content for the file.",
+            example = "Some text.")
+    private final String fileContents;
+
+    @Option(displayName = "Relative File Path",
+            description = "File path of new file.",
+            example = "foo/bar/baz.txt")
+    private final String relativeFileName;
+
+    @Option(displayName = "Overwrite existing file",
+            description = "If there is an existing file, should it be overwritten.")
+    private final Boolean overwriteExisting;
+
+    public CreateTextFile(String fileContents, String relativeFileName, Boolean overwriteExisting) {
+        this.fileContents = fileContents;
+        this.relativeFileName = relativeFileName;
+        this.overwriteExisting = overwriteExisting;
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("plain text");
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Create text file";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Creates a new plain text file.";
+    }
+
+
+    @Override
+    protected List<SourceFile> visit(List<SourceFile> before, ExecutionContext ctx) {
+        Path path = Paths.get(relativeFileName);
+        List<SourceFile> results = new ArrayList<>(before);
+        UUID matchingFileId = null;
+
+        Optional<SourceFile> matchingFile = results.stream().filter(sourceFile -> path.toString().equals(sourceFile.getSourcePath().toString())).findFirst();
+
+        if (matchingFile.isPresent()) {
+            matchingFileId = matchingFile.get().getId();
+        }
+
+        Parser.Input newFile = new Parser.Input(
+                path,
+                () -> new ByteArrayInputStream(fileContents.getBytes(StandardCharsets.UTF_8)),
+                true
+        );
+
+        List<PlainText> newFiles = new PlainTextParser().parseInputs(Arrays.asList(newFile), null, ctx);
+        PlainText brandNewFile = newFiles.get(0);
+
+        if (matchingFileId != null && overwriteExisting) {
+            brandNewFile = new PlainText(matchingFileId, brandNewFile.getSourcePath(), brandNewFile.getMarkers(), brandNewFile.getText());
+            results.add(brandNewFile);
+        } else if (matchingFileId == null) {
+            results.add(brandNewFile);
+        } else {
+            return before;
+        }
```

Comment:
It looks like `matchingFileId` and `overwriteExisting` are known before we bother creating any text files or parsers. Might as well return as soon as these are known without spending time creating objects that don't get used.

Reply:
makes total sense. went back and tried to simplify the logic and I think it reads clearer now